### PR TITLE
SNS create_topic should only accept valid characters

### LINKED ISSUE #1328: SNS.create_topic shouldn't accept invalid input
Per the [AWS SNS documentation](http://docs.aws.amazon.com/sns/latest/api/API_CreateTopic.html) the CreateTopic shouldn't accept topic names that don't match the constraint: _Topic names must be made up of only uppercase and lowercase ASCII letters, numbers, underscores, and hyphens, and must be between 1 and 256 characters long._

Expected:
- create_topic should raise an InvalidParameterValue when an invalid topic is passed to create_topic

Results:
- no error is raised

### PR DESCRIPTION
**Pull request by adamgilman**
Closes #1328

### CODE REVIEW

**Comment by adamgilman** [comment]
Fixing Travis’ whining now

**Comment by terrycain** [comment]
Ha yeah flake8 gets me alot

**Comment by terrycain** [comment]
The API docs also don't mention spaces, the regex will need changing
If were going on the assumption that we don't want a topic to start or end with a _/- this regex will do that and enforce the length is at least 1 and not longer than 256 too `^[a-zA-Z0-9](?:[A-Za-z0-9_-]{0,253}[a-zA-Z0-9])?$`

File: moto/sns/models.py [hunk 1]
```diff
@@ -193,6 +195,9 @@ def update_sms_attributes(self, attrs):
         self.sms_attributes.update(attrs)
 
     def create_topic(self, name):
+        fails_constraints = not re.match(r'^[a-zA-Z0-9](?:[A-Za-z0-9_-]{0,253}[a-zA-Z0-9])?$', name)
+        if fails_constraints:
+            raise InvalidParameterValue("Can only include alphanumeric characters, hyphens, or underscores. 1 to 80 in length")
```
Comment (terrycain) [inline moto/sns/models.py]: 1 to 256 ?

Comment (adamgilman) [reply moto/sns/models.py]: Just copy and pasted what you gave me :)

Comment (terrycain) [reply moto/sns/models.py]: I meant the string in the exception

**Comment by JackDanger** [comment]
@adamgilman this is just about finished, there are just a couple of comments by Terry that need to be addressed. Thanks for doing this!

**Comment by adamgilman** [comment]
Sorry for the delay. Been busy at work and have it patched locally. Will
try to get a PR done today

**Comment by JackDanger** [comment]
Nice! @terrycain please merge if/when you approve this.

**Comment by terrycain** [comment]
Yeah looks gread :smile:

**Pull request merged**